src/redprl/refiner.fun

```diff
@@ -498,80 +498,48 @@ struct
         |>: goal #> (I, H, hole)
       end
 
-    fun HeadExpansionDelegate sign (I, H) jdgMaker m =
-      let
-        val (goal, hole) = makeGoal @@ jdgMaker @@
-          Machine.eval sign Machine.CUBICAL (Machine.Unfolding.default sign) m
-      in
-        |>: goal #> (I, H, hole)
-      end
+    fun eval sign =
+      Machine.eval sign Machine.CUBICAL (Machine.Unfolding.default sign)
 
     (* favonia: the following should be generalized and put into the Sequent
      * module such that a single template function `HeadExpansionDelegate`
      * can generate all the following functions.
      *)
 
-    fun EqHeadExpansionL sign _ jdg =
-      let
-        val _ = RedPrlLog.trace "Computation.EqHeadExpansionL"
-        val           (I, H) >> CJ.EQ ((m, n), (ty, l, k)) = jdg
-        fun maker m = (I, H) >> CJ.EQ ((m, n), (ty, l, k))
-      in
-        HeadExpansionDelegate sign (I, H) maker m
-      end
-
-    fun EqHeadExpansionR sign _ jdg =
-      let
-        val _ = RedPrlLog.trace "Computation.EqHeadExpansionR"
-        val           (I, H) >> CJ.EQ ((m, n), (ty, l, k)) = jdg
-        fun maker n = (I, H) >> CJ.EQ ((m, n), (ty, l, k))
-      in
-        HeadExpansionDelegate sign (I, H) maker n
-      end
-
-    fun EqHeadExpansionTy sign _ jdg =
-      let
-        val _ = RedPrlLog.trace "Computation.EqHeadExpansionTy"
-        val            (I, H) >> CJ.EQ ((m, n), (ty, l, k)) = jdg
-        fun maker ty = (I, H) >> CJ.EQ ((m, n), (ty, l, k))
-      in
-        HeadExpansionDelegate sign (I, H) maker ty
-      end
-
-    fun EqTypeHeadExpansionL sign _ jdg =
+    fun CatJdgHeadExpansion sign _ jdg =
       let
-        val _ = RedPrlLog.trace "Computation.EqTypeHeadExpansionL"
-        val           (I, H) >> CJ.EQ_TYPE ((a, b), l, k) = jdg
-        fun maker a = (I, H) >> CJ.EQ_TYPE ((a, b), l, k)
+        val _ = RedPrlLog.trace "Computation.CatJdgHeadExpansion"
+        val (I, H) >> catjdg = jdg
+        val (goal, hole) = makeGoal @@ (I, H) >> CJ.map (eval sign) catjdg
       in
-        HeadExpansionDelegate sign (I, H) maker a
+        |>: goal #> (I, H, hole)
       end
 
-    fun EqTypeHeadExpansionR sign _ jdg =
+    fun SequentHeadExpansion sign _ jdg =
       let
-        val _ = RedPrlLog.trace "Computation.EqTypeHeadExpansionR"
-        val           (I, H) >> CJ.EQ_TYPE ((a, b), l, k) = jdg
-        fun maker b = (I, H) >> CJ.EQ_TYPE ((a, b), l, k)
+        val _ = RedPrlLog.trace "Computation.SequentHeadExpansion"
+        val (I, H) >> catjdg = jdg
+        val (goal, hole) = makeGoal @@ Seq.map (eval sign) jdg
       in
-        HeadExpansionDelegate sign (I, H) maker b
+        |>: goal #> (I, H, hole)
       end
 
-    fun TrueHeadExpansion sign _ jdg =
+    fun MatchHeadExpansion sign _ jdg =
```

Comment:
I think we should be cautious about having head expansion for the "match" judgments. I'm not completely sure of what there semantics would be if we have such a rule.

Reply:
Hmm... are you worried about some computation turning something matched into something unmatched?

Reply:
Anyways I removed it.

Reply:
Yeah, something like that... It might be OK, but it's hard for me to say what the meaning of the judgment is if we have this rule—I envisioned it as a very syntactic thing.

Preferably, this computation would be done *before* the match subgoal was generated (somehow).